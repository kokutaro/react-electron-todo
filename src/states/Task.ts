export interface Task {
  complete: boolean;
  deadline: Date;
  id: string;
  taskName: string;
}

export interface TaskList {
  tasks: Task[];
}
