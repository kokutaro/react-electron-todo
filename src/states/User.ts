/**
 * User definition
 */

export default interface User {
  /** Name */
  name: string;

  /** Count */
  count: number;
}
