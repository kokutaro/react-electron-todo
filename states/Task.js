"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiVGFzay5qcyIsInNvdXJjZVJvb3QiOiIuL3NyYy8iLCJzb3VyY2VzIjpbInN0YXRlcy9UYXNrLnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiIiLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgaW50ZXJmYWNlIFRhc2sge1xuICBjb21wbGV0ZTogYm9vbGVhbjtcbiAgZGVhZGxpbmU6IERhdGU7XG4gIGlkOiBzdHJpbmc7XG4gIHRhc2tOYW1lOiBzdHJpbmc7XG59XG5cbmV4cG9ydCBpbnRlcmZhY2UgVGFza0xpc3Qge1xuICAvKiogTGlzdCBvZiB0YXNrcyAqL1xuICB0YXNrczogVGFza1tdO1xuICAvKiogQ2hlY2sgaWYgdGFzayBpcyBsb2FkaW5nICovXG4gIGxvYWRpbmc6IGJvb2xlYW47XG4gIC8qKiBGYWlsdXJlIG1lc3NhZ2UgKi9cbiAgZmFpbHVyZU1lc3NhZ2U6IHN0cmluZztcbn1cbiJdfQ==