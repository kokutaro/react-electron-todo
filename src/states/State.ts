import { TaskList } from './Task';

export interface State {
  taskList: TaskList;
}
