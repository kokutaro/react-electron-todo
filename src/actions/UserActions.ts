import { actionCreatorFactory } from 'typescript-fsa';

import User from '../states/User';
const actionCreator = actionCreatorFactory('USER_ACTION');

export const changeUserAction = actionCreator<Partial<User>>('CHANGE_USER');
