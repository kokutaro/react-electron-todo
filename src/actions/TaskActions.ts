import '../core/CoreInterface';

import { Dispatch } from 'redux';
import { actionCreatorFactory } from 'typescript-fsa';

import { Task } from '../states/Task';

const actionCreator = actionCreatorFactory('TASK_ACTIONS');

export const showTaskListAction = actionCreator.async<null, Task[], string>('SHOW_TASK_LIST');

export const getTaskList = async (dispatch: Dispatch): Promise<void> => {
  dispatch(showTaskListAction.started(null));
  const taskList = await window.core.loadTaskList().catch((e) => {
    console.error(e);
    dispatch(
      showTaskListAction.failed({
        error: e.toString(),
        params: null,
      }),
    );
  });
  if (!taskList) {
    return;
  }
  dispatch(showTaskListAction.done({ result: taskList, params: null }));
};

export const addTask = async (task: Task, dispatch: Dispatch): Promise<void> => {
  dispatch(showTaskListAction.started(null));
  const taskList = await window.core.saveTask(task).catch((e) => {
    console.error(e);
    dispatch(
      showTaskListAction.failed({
        error: e.toString(),
        params: null,
      }),
    );
  });
  if (!taskList) {
    return;
  }
  dispatch(
    showTaskListAction.done({
      result: taskList,
      params: null,
    }),
  );
};

export const toggleTask = async (task: Task, dispatch: Dispatch): Promise<void> => {
  dispatch(showTaskListAction.started(null));
  task.complete = !task.complete;
  const taskList = await window.core.saveTask(task).catch((e) => {
    console.error(e);
    dispatch(
      showTaskListAction.failed({
        error: e.toString(),
        params: null,
      }),
    );
  });
  if (!taskList) {
    return;
  }
  dispatch(
    showTaskListAction.done({
      result: taskList,
      params: null,
    }),
  );
};

export const deleteTask = async (taskId: string, dispatch: Dispatch): Promise<void> => {
  dispatch(showTaskListAction.started(null));
  const taskList = await window.core.deleteTask(taskId).catch((e) => {
    console.error(e);
    dispatch(
      showTaskListAction.failed({
        error: e.toString(),
        params: null,
      }),
    );
  });
  if (!taskList) {
    return;
  }
  dispatch(
    showTaskListAction.done({
      result: taskList,
      params: null,
    }),
  );
};
