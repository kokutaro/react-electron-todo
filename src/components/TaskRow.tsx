import moment from 'moment';
import React, { MouseEvent, useCallback, useMemo } from 'react';
import { useDispatch } from 'react-redux';

import { deleteTask, toggleTask } from '../actions/TaskActions';
import { Task } from '../states/Task';
import { styled } from './FoundationStyles';

const TaskCell = styled.div<{ expiration: boolean }>`
  align-items: center;
  background-color: ${(p): string => (p.expiration ? 'inherit' : p.theme.SECONDARY_2_0)};
  border: 1px solid rgb(200, 200, 200);
  border-radius: 5px;
  cursor: pointer;
  display: flex;
  flex-direction: row;
  margin-bottom: 1em;
  padding: 10px;
  transition-duration: 0.2s;
  transition-property: all;

  &:hover {
    box-shadow: 5px 5px 5px rgba(200, 200, 200, 4);
    transform: translate(-5px, -5px);
  }
`;

const TaskCheckBox = styled.div`
  align-items: center;
  background-color: #fff;
  border: 2px solid #ccc;
  border-radius: 50%;
  display: flex;
  flex-grow: 0;
  flex-shrink: 0;
  height: 2em;
  justify-content: center;
  width: 2em;
`;

const TaskCheck = styled.p`
  color: ${(p): string => p.theme.SECONDARY_1_3};
  font-size: 150%;
`;

const TaskBody = styled.div`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  flex-shrink: 0;
  height: 3em;
  justify-content: space-around;
`;

const TaskRemove = styled.div`
  flex-grow: 0;
  flex-shrink: 0;
`;

const TaskName = styled.div`
  font-size: 120%;
`;

const DeadLine = styled.div``;

const TaskRow: React.FC<{ data: Task }> = (props: { data: Task }) => {
  const { data } = props;
  const dispatch = useDispatch();

  const expiration = useMemo(() => {
    return new Date() < data.deadline || data.complete;
  }, [data.complete, data.deadline]);

  const deadLineString = useMemo(() => {
    return moment(data.deadline).format('L LT');
  }, [data.deadline]);

  const onRowClick = useCallback(() => {
    toggleTask(data, dispatch);
  }, [data.id]);

  const onDeleteClick = useCallback(
    (e: MouseEvent<HTMLDivElement>) => {
      deleteTask(data.id, dispatch);
      e.stopPropagation();
    },
    [data.id],
  );

  return (
    <TaskCell className="taskCell" expiration={expiration} onClick={onRowClick}>
      <TaskCheckBox>
        <TaskCheck>{data.complete ? '✔' : null}</TaskCheck>
      </TaskCheckBox>
      <TaskBody>
        <TaskName>{data.taskName}</TaskName>
        <DeadLine>⏰{deadLineString}</DeadLine>
      </TaskBody>
      <TaskRemove className="deleteButton" onClick={onDeleteClick}>
        ❌
      </TaskRemove>
    </TaskCell>
  );
};

export default TaskRow;
