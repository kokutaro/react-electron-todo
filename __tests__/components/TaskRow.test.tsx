/**
 * @jest-environment jsdom
 */

/* eslint-disable @typescript-eslint/explicit-function-return-type */

import 'jest-styled-components';
import 'ts-jest';

import enzyme, { mount } from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import React from 'react';

import { deleteTask, toggleTask } from '../../src/actions/TaskActions';
import TaskRow from '../../src/components/TaskRow';
import { Task } from '../../src/states/Task';

beforeAll(() => {
  enzyme.configure({
    adapter: new EnzymeAdapter(),
    disableLifecycleMethods: true,
  });
});

const dispatch = jest.fn();

jest.mock('react-redux', () => ({
  useDispatch: () => dispatch,
}));

jest.mock('../../src/actions/TaskActions');

describe('TaskRow', () => {
  const task1: Task = {
    complete: true,
    deadline: new Date('2020-05-01T21:24:00.000Z'),
    id: 'task001',
    taskName: 'task 001',
  };

  const task2: Task = {
    complete: false,
    deadline: new Date('2020-05-01T21:24:00.000Z'),
    id: 'task002',
    taskName: 'task 002',
  };

  test('complete', () => {
    const wrapper = mount(<TaskRow data={task1} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('incomplete over due', () => {
    const wrapper = mount(<TaskRow data={task2} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('click delete', () => {
    const wrapper = mount(<TaskRow data={task1} />);
    const deleteButton = wrapper.find('div.deleteButton');
    deleteButton.simulate('click');
    expect(deleteTask).toBeCalledWith('task001', dispatch);
  });

  test('click complete', () => {
    const wrapper = mount(<TaskRow data={task1} />);
    const deleteButton = wrapper.find('div.taskCell');
    deleteButton.simulate('click');
    expect(toggleTask).toBeCalledWith(task1, dispatch);
  });
});
