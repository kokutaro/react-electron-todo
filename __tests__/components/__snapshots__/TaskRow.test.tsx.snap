// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskRow complete 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: inherit;
  border: 1px solid rgb(200,200,200);
  border-radius: 5px;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-bottom: 1em;
  padding: 10px;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  -webkit-transition-property: all;
  transition-property: all;
}

.c0:hover {
  box-shadow: 5px 5px 5px rgba(200,200,200,4);
  -webkit-transform: translate(-5px,-5px);
  -ms-transform: translate(-5px,-5px);
  transform: translate(-5px,-5px);
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #fff;
  border: 2px solid #ccc;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 2em;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 2em;
}

.c2 {
  font-size: 150%;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 3em;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c5 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 {
  font-size: 120%;
}

<TaskRow
  data={
    Object {
      "complete": true,
      "deadline": 2020-05-01T21:24:00.000Z,
      "id": "task001",
      "taskName": "task 001",
    }
  }
>
  <styled.div
    className="taskCell"
    expiration={true}
    onClick={[Function]}
  >
    <div
      className="c0 taskCell"
      onClick={[Function]}
    >
      <styled.div>
        <div
          className="c1"
        >
          <styled.p>
            <p
              className="c2"
            >
              ✔
            </p>
          </styled.p>
        </div>
      </styled.div>
      <styled.div>
        <div
          className="c3"
        >
          <styled.div>
            <div
              className="c4"
            >
              task 001
            </div>
          </styled.div>
          <styled.div>
            <div
              className=""
            >
              ⏰
              05/01/2020 2:24 PM
            </div>
          </styled.div>
        </div>
      </styled.div>
      <styled.div
        className="deleteButton"
        onClick={[Function]}
      >
        <div
          className="c5 deleteButton"
          onClick={[Function]}
        >
          ❌
        </div>
      </styled.div>
    </div>
  </styled.div>
</TaskRow>
`;

exports[`TaskRow incomplete over due 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid rgb(200,200,200);
  border-radius: 5px;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-bottom: 1em;
  padding: 10px;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  -webkit-transition-property: all;
  transition-property: all;
}

.c0:hover {
  box-shadow: 5px 5px 5px rgba(200,200,200,4);
  -webkit-transform: translate(-5px,-5px);
  -ms-transform: translate(-5px,-5px);
  transform: translate(-5px,-5px);
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #fff;
  border: 2px solid #ccc;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 2em;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 2em;
}

.c2 {
  font-size: 150%;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 3em;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c5 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 {
  font-size: 120%;
}

<TaskRow
  data={
    Object {
      "complete": false,
      "deadline": 2020-05-01T21:24:00.000Z,
      "id": "task002",
      "taskName": "task 002",
    }
  }
>
  <styled.div
    className="taskCell"
    expiration={false}
    onClick={[Function]}
  >
    <div
      className="c0 taskCell"
      onClick={[Function]}
    >
      <styled.div>
        <div
          className="c1"
        >
          <styled.p>
            <p
              className="c2"
            />
          </styled.p>
        </div>
      </styled.div>
      <styled.div>
        <div
          className="c3"
        >
          <styled.div>
            <div
              className="c4"
            >
              task 002
            </div>
          </styled.div>
          <styled.div>
            <div
              className=""
            >
              ⏰
              05/01/2020 2:24 PM
            </div>
          </styled.div>
        </div>
      </styled.div>
      <styled.div
        className="deleteButton"
        onClick={[Function]}
      >
        <div
          className="c5 deleteButton"
          onClick={[Function]}
        >
          ❌
        </div>
      </styled.div>
    </div>
  </styled.div>
</TaskRow>
`;
